### ISSUE TITLE
Rule proposal: no-unsafe-type-assertion

### ISSUE THREAD

**Issue by lgenzelis** [typescript-eslint/typescript-eslint#7173]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

One of the hardest thing to explain to people not experienced in TS is when it's ok to use `as`. I default to tell people "don't use `as`", and I constantly have to deal with that in code reviews. I'd love to have a rule that only allows type assertions that widen a type. For example, `as unknown` is always totally ok. `as string | number` is safe on a `string`, but unsafe in a `string | boolean`.

This rule should only allow doing `x as Type` in the cases where the type of `x` is assignable to `type`. So, the rule of thumb would be: if `const a: Type = x` is fine, then `x as Type` is fine. Otherwise it's not.

### Fail Cases

```typescript
const x = 'something';
const y = x as unknown as number; // Type `unknown` is not assignable to type `number` 🚫

//-----------

function f() {
  return Math.random() < 0.5 ? 42 : 'meh';
}
const z = f() as number; // Type `number | string` is not assignable to type `number` 🚫
```

### Pass Cases

```typescript
const a = 'something';
const b = a as unknown; // Type `string` is assignable to type `unknown` ✅

//-----------

function g() {
  return Math.random() < 0.5 ? 42 : 'meh';
}
const h = f() as number | string | boolean; // Type `number | string` is assignable to type `number | string | boolean` ✅

//-----------

const person = {
    name: 'Chuck',
    lastName: 'Norris' as string | undefined // Type `Norris` is assignable to type `string | undefined`  ✅
}
person.lastName = undefined; // this is fine because of the type assertion (just showing why we'd want the `as string | undefined`)
```

### Additional Info

Here's a[ link to a TS playground](https://www.typescriptlang.org/play?#code/PQKgtBldkAQCoAsCmBnZtWIPYFcA2AJrAOYCWAbhgIYB2syATo9o7DB9CMAFA8DG2WqgAusAB6wAvLADkqbAFtkIxGVolZAbgFDRsAJ7SJsaqli5aAa1rYA7vTOxauRQCMmW2MGAIDABwwAA0sbe1og2DJzWzEzVDISWmo3fAwRbFgRAOCXdyZIwF4NwGq9vh9OSD4AM0t+ETIhWCqACgBKWABvHlhYRhVcRnoAWWpVADpGOkIlNtgAHlgABjGAVlgAflgAFgAmWAAuOWVEbR4AX11hMQAvYxb2pzyPRi8fP0DYIKemWAAfTBEjHUJEi0Wc2DiqASSRSaSymWyHy+rmeRVKPFAFQqCBQ6FMfSa6gwWM43D4giupmM8iUKjUGlOFP0bmM1FM5lCtgcr188Byn1EQI0oPM8USyVS6QR-JC1i5EVggFByMrAEnVWr1RokWZdHp9EQDYajRATKYzdoLZZrTa7A5HZAnHQXJliRB3WaPFE-f6C4F-WBubDYNJ0HnvXJetg+wHAkXs6ESuEZLIy75RgFCkj+wPB5B0SLKjGqrHkvRiQKMBT0GS6nrOajKQ6yADCiFw-CssgANN06-gzCIAHIN5BNwesIGoWTsjN+-6WQjIKpE4hvPlI8fMaJxsUwyXwlNI30af0LpcryJK848CtVsb90TD5TGM-L2jIQhh1RgsFvjAefhqFwPFsCqLIUEPGgoSYTV6GaAArYCxCwew-TsRAjDsZBZGIOw6DEVRgicY8s3nWhFz-QgglaHggA) with the passing and (proposed) failing cases.

**Comment by JoshuaKGoldberg** [comment]
I bet this would be a fascinating rule to figure out the details of... but it's blocked by the lack of a TypeScript _assignability_ API: https://github.com/microsoft/TypeScript/issues/9879. 

Related:
* #709: not the same request, but in the same area of blocked functionality
* #3226: somewhat a dup, but that's an old issue from before our `blocked by external API` label and has a less flavorful OP

**Comment by JoshuaKGoldberg** [comment]
I bet this would be a fascinating rule to figure out the details of... but it's blocked by the lack of a TypeScript _assignability_ API: https://github.com/microsoft/TypeScript/issues/9879. 

As much as I love a good Chuck Norris reference, not much point to discussing here until we have a TypeScript assignability API. 😞 

Related:
* #709: not the same request, but in the same area of blocked functionality
* #3226: somewhat a dup, but that's an old issue from before our `blocked by external API` label and has a less flavorful OP

**Comment by lgenzelis** [comment]
Wow, I didn't know such an API didn't exist. I can't imagine how hard it must be to write rules without having a tool as basic as `isTypeAssignableTo` :/ 

In the meantime, for my use case, I can just use a noop function that takes care of this. Something like:
```typescript
function safeAs<T>(x: T) {
  return x;
}
```
This allows you to do `safeAs<string | undefined>('some text')`, but prevents you from doing 
```
const x: string | number = 42;
const y = safeAs<string>(x);
```

[Playground link here](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAZwIbAKYEFkB4AqAfABQAeAXIvgJSIDeAUIogE4ZQgtKkDcDAvgwYQEyKIlSIAvCnTY8YljDABzRAB9E4ACYZgyjNpIByZHAC2GRFAykox6n2GjxFFFCWqNiMCHMAjDBZpRAAWACY+ETAxRABPELRMHFxFZRUSUkchIA) :)

Anyways, thanks for your quick response @JoshuaKGoldberg ! I leave the decision up to you as to whether it makes sense or not to keep this issue open.

**Comment by bradzacher** [comment]
The `satisfies` operator will do the same check as a no-op function (but it doesn't change the type). A variable annotation also does the same safe checks (and it does change the type).

**Comment by lgenzelis** [comment]
Thanks @bradzacher ! The issue with a variable annotation is that it requires a separate statement. I believe the place where I use `as` the most is in large objects in files with mock data for tests. I want to be able to have an object like
```typescript
const bigObject = {
  name: 'Random',
  lastName: 'Guy',
  age: 42,
}
```

and then, in one test, do 
```typescript
bigObject.age = '42'; // BE sometimes sends age as a string
```

Since the objects are really large, I don't want to write the types myself. So, my "only" option is writing
```typescript
const bigObject = {
  name: 'Random',
  lastName: 'Guy',
  age: 42 as string | number,
}
```
To use a variable annotation I would need to declare `const age: string | number` beforehand. And since I have many objects (and many properties which require these kind of annotations), doing that would be cumbersome 😅

**Comment by TSMMark** [comment]
I would love to see a eslint rule to disallow `as` entirely, or prefer `satisfies` over `as` — or something along those lines.

Does such a rule exist? I searched around but couldn't find it. Would the proposal in this issue cover that? Or would maintainers prefer I open a separate issue?

**Comment by JoshuaKGoldberg** [comment]
https://github.com/typescript-eslint/typescript-eslint/issues/709

**Comment by TSMMark** [comment]
> #709 😢 blocked on [microsoft/TypeScript#9879](https://github.com/microsoft/TypeScript/issues/9879)

Thansk @JoshuaKGoldberg 

I might take a custom linter into my own hands then. Do you foresee any issues if I wrote myself a custom eslint rule, which would simply look for the `as` operator and report an offense in 100% of cases?

In this codebase specifically, I'm okay saying that "if the author really knows what they're doing and needs to use `as`, they can eslint-disable directive comment for that line"

**Comment by bradzacher** [comment]
@TSMMark for disallowing `as` see https://typescript-eslint.io/rules/consistent-type-assertions

**Comment by controversial** [comment]
Note Typescript 5.4 made `isTypeAssignableTo` public:
- https://github.com/microsoft/TypeScript/pull/56448
- https://github.com/microsoft/TypeScript/issues/9879#issuecomment-1847859292
 
@JoshuaKGoldberg created a discussion stating that it will be usable in the next major: https://github.com/typescript-eslint/typescript-eslint/discussions/7936

Does this mean a `no-unsafe-type-assertion` rule will be possible to add to typescript-eslint v8?

**Comment by JoshuaKGoldberg** [comment]
That should be right, yup! 🙌

**Comment by kirkwaiblinger** [comment]
🎉  For anyone that comes along after v8 is released, consider this "accepting PRs" (and please tag one of us to poke us so we can change the label)

**Comment by JoshuaKGoldberg** [comment]
Now that v8 is released, per https://github.com/typescript-eslint/typescript-eslint/discussions/7936, we can mark this issue as unblocked! 🚀

**Comment by Samuel-Therrien-Beslogic** [comment]
I should mention, if anyone's looking for inspiration, there's this implementation by a third-party plugin: https://github.com/danielnixon/eslint-plugin-total-functions/blob/master/src/rules/no-unsafe-type-assertion.ts

Also, it'd be nice to properly document and teach alternatives to unsafe type-casts, here's the note I had in our internal linting docs:
```js
// It takes an advanced TypeScript user to fully make sense of this rule
/**
 * It can sometimes be hard to make sense of this rule, but let's be strict!
 * Tips:
 * - Prefer coercion over assertion: e.g `String(value ?? '')` over `value as string | null`
 *   (this depends on your functional logic)
 * - Try to assign to an explicitly typed variable and TypeScript will tell you about the type difference!
 *     ie: `const newValue: ExpectedType = unsafeValue`
 * - If using TypeScript 4.9+, consider using the `satisfies` Operator:
 *     (https://devblogs.microsoft.com/typescript/announcing-typescript-4-9/#the-satisfies-operator)
 *     `const newValue = {a: "A", b: 2} satisfies ExpectedType`
 * - If using TypeScript 4.9+, consider using the `in` Operator to narrow unlisted properties:
 *     (https://devblogs.microsoft.com/typescript/announcing-typescript-4-9/#unlisted-property-narrowing-with-the-in-operator)
 *     `if ("name" in unsafeValue && typeof unsafeValue.name === "string") ...`
 * - Sometimes just casting as `Partial<>` is sufficient
 * - Double check that you are not accidentally removing `null | undefined` from the type when casting:
 *   - `inputEvent.target as Partial<HTMLInputElement>` <-- Oops!
 *   - `event.target as Partial<HTMLInputElement> | null` <-- Ok!
 */
```

**Comment by Guziq** [comment]
Thank you for your message. I am currently on holiday until the end of November. For any urgent matters, please feel free to reach out to my team at ***@***.*** you for your understanding, and I will get back to you as soon as possible upon my return.Best regards,Jakub
